File: app/src/main/java/com/commit451/gitlab/fragment/FeedFragment.java
package com.commit451.gitlab.fragment;

import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.commit451.gitlab.App;
import com.commit451.gitlab.R;
import com.commit451.gitlab.adapter.DividerItemDecoration;
import com.commit451.gitlab.adapter.FeedAdapter;
import com.commit451.gitlab.model.rss.Entry;
import com.commit451.gitlab.model.rss.Feed;
import com.commit451.gitlab.navigation.Navigator;
import com.commit451.gitlab.rx.CustomSingleObserver;
import com.novoda.simplechromecustomtabs.SimpleChromeCustomTabs;

import butterknife.BindView;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

/**
 * Takes an RSS feed url and shows the feed
 */
public class FeedFragment extends ButterKnifeFragment {

    private static final String EXTRA_FEED_URL = "extra_feed_url";

    public static FeedFragment newInstance(Uri feedUrl) {
        Bundle args = new Bundle();
        args.putParcelable(EXTRA_FEED_URL, feedUrl);

        FeedFragment fragment = new FeedFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @BindView(R.id.swipe_layout)
    SwipeRefreshLayout swipeRefreshLayout;
    @BindView(R.id.list)
    RecyclerView listEntries;
    @BindView(R.id.message_text)
    TextView textMessage;

    private Uri feedUrl;
    private FeedAdapter adapterFeed;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        feedUrl = getArguments().getParcelable(EXTRA_FEED_URL);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_feed, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        adapterFeed = new FeedAdapter(new FeedAdapter.Listener() {
            @Override
            public void onFeedEntryClicked(Entry entry) {
                Navigator.navigateToUrl(getActivity(), entry.getLink().getHref(), App.get().getAccount());
            }
        });
        listEntries.setLayoutManager(new LinearLayoutManager(getActivity()));
        listEntries.addItemDecoration(new DividerItemDecoration(getActivity()));
        listEntries.setAdapter(adapterFeed);

        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                loadData();
            }
        });

        loadData();
    }

    @Override
    public void onResume() {
        super.onResume();
        SimpleChromeCustomTabs.getInstance().connectTo(getActivity());
    }

    @Override
    public void onPause() {
        if (SimpleChromeCustomTabs.getInstance().isConnected()) {
            SimpleChromeCustomTabs.getInstance().disconnectFrom(getActivity());
        }
        super.onPause();
    }

    @Override
    protected void loadData() {
        if (getView() == null) {
            return;
        }
        if (feedUrl == null) {
            swipeRefreshLayout.setRefreshing(false);
            textMessage.setVisibility(View.VISIBLE);
            return;
        }
        textMessage.setVisibility(View.GONE);
        swipeRefreshLayout.post(new Runnable() {
            @Override
            public void run() {
                if (swipeRefreshLayout != null) {
                    swipeRefreshLayout.setRefreshing(true);
                }
            }
        });
        App.get().getGitLabRss().getFeed(feedUrl.toString())
                .compose(this.<Feed>bindToLifecycle())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new CustomSingleObserver<Feed>() {
                    @Override
                    public void success(@NonNull Feed feed) {
                        swipeRefreshLayout.setRefreshing(false);
                        if (feed.getEntries() != null && !feed.getEntries().isEmpty()) {
                            textMessage.setVisibility(View.GONE);
                        } else {
                            Timber.d("No activity in the feed");
                            textMessage.setVisibility(View.VISIBLE);
                            textMessage.setText(R.string.no_activity);
                        }
                        adapterFeed.setEntries(feed.getEntries());
                    }

                    @Override
                    public void error(@NonNull Throwable e) {
                        Timber.e(e);
                        swipeRefreshLayout.setRefreshing(false);
                        textMessage.setVisibility(View.VISIBLE);
                        textMessage.setText(R.string.connection_error_feed);
                        adapterFeed.setEntries(null);
                    }
                });
    }
}

File: app/src/main/java/com/commit451/gitlab/fragment/FeedFragment.kt
package com.commit451.gitlab.fragment

import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.commit451.addendum.design.snackbar
import com.commit451.gitlab.App
import com.commit451.gitlab.R
import com.commit451.gitlab.adapter.BaseAdapter
import com.commit451.gitlab.model.rss.Entry
import com.commit451.gitlab.navigation.Navigator
import com.commit451.gitlab.util.LoadHelper
import com.commit451.gitlab.viewHolder.FeedEntryViewHolder
import com.novoda.simplechromecustomtabs.SimpleChromeCustomTabs
import kotlinx.android.synthetic.main.fragment_feed.*
import kotlinx.android.synthetic.main.fragment_feed.swipeRefreshLayout

/**
 * Takes an RSS feed url and shows the feed
 */
class FeedFragment : BaseFragment() {

    companion object {

        private const val EXTRA_FEED_URL = "extra_feed_url"

        fun newInstance(feedUrl: String): FeedFragment {
            val args = Bundle()
            args.putString(EXTRA_FEED_URL, feedUrl)

            val fragment = FeedFragment()
            fragment.arguments = args
            return fragment
        }
    }

    private lateinit var adapter: BaseAdapter<Entry, FeedEntryViewHolder>
    private lateinit var loadHelper: LoadHelper<Entry>

    private var feedUrl: Uri? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        feedUrl = Uri.parse(arguments?.getString(EXTRA_FEED_URL))
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_feed, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        adapter = BaseAdapter(
                onCreateViewHolder = { parent, _ ->
                    val viewHolder = FeedEntryViewHolder.inflate(parent)
                    viewHolder.itemView.setOnClickListener {
                        val entry = adapter.items[viewHolder.adapterPosition]
                        if (entry.link.href.isEmpty()) {
                            root.snackbar(R.string.not_a_valid_url)
                        } else {
                            Navigator.navigateToUrl(baseActivty, entry.link.href, App.get().getAccount())
                        }
                    }
                    viewHolder
                },
                onBindViewHolder = { viewHolder, _, item -> viewHolder.bind(item) }
        )
        loadHelper = LoadHelper(
                lifecycleOwner = this,
                recyclerView = listEntries,
                baseAdapter = adapter,
                swipeRefreshLayout = swipeRefreshLayout,
                dividers = true,
                errorOrEmptyTextView = textMessage,
                loadInitial = {
                    gitLab.feed(feedUrl!!.toString())
                }
        )

        loadData()
    }

    override fun onResume() {
        super.onResume()
        SimpleChromeCustomTabs.getInstance().connectTo(baseActivty)
    }

    override fun onPause() {
        if (SimpleChromeCustomTabs.getInstance().isConnected) {
            SimpleChromeCustomTabs.getInstance().disconnectFrom(baseActivty)
        }
        super.onPause()
    }

    override fun loadData() {
        loadHelper.load()
    }
}
